src/test/pybind/test_rados.py

```diff
@@ -598,6 +598,24 @@ def test_remove_omap_ramge2(self):
             self.ioctx.operate_read_op(read_op, "test_obj")
             eq(list(iter), [("4", b"dddd")])
 
+    def test_cmpext_op(self):
+        with WriteOpCtx() as write_op:
+            write_op.write(b'12345', 0)
+            self.ioctx.operate_write_op(write_op, "test")
+            eq(self.ioctx.read("test"), b'12345')
+            write_op.cmpext(b'12345', 0)
+            self.ioctx.operate_write_op(write_op, "test")
```

Comment:
@tchaikov   Thanks a lot for the suggestions
```
@param prval returned result of comparison, 0 on success, negative error code on failure, (-MAX_ERRNO - mismatch_off) on mismatch
CEPH_RADOS_API void rados_write_op_cmpext(rados_write_op_t write_op,
                                          const char *cmp_buf,
                                          size_t cmp_len,
                                          uint64_t off,
                                          int *prval);


def cmpext(self, cmp_buf: bytes, offset: int = 0):
        """
        Ensure that given object range (extent) satisfies comparison
        :param cmp_buf: buffer containing bytes to be compared with object contents
        :param offset: object byte offset at which to start the comparison
        """

        cdef:
            char *_cmp_buf = cmp_buf
            size_t _cmp_buf_len = len(cmp_buf)
            uint64_t _offset = offset
        with nogil:
            rados_write_op_cmpext(self.write_op, _cmp_buf, _cmp_buf_len, _offset, NULL)
```
Cannot get error number because prval = NULL

Reply:
```
import rbd
from rados import (Rados, WriteOpCtx, ReadOpCtx, ObjectNotFound, OSError)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import rados
import unittest
import logging

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool_01')
        assert self.rados.pool_exists('pool_01')
        self.ioctx = self.rados.open_ioctx('pool_01')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool_01')
        self.rados.shutdown()

    def test_cmpext_op(self):
        object_id = 'test'
        with WriteOpCtx() as write_op:
            write_op.write(b'12345', 0)
            self.ioctx.operate_write_op(write_op, object_id)
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'12345', 0)
            write_op.write(b'54321', 0)
            self.ioctx.operate_write_op(write_op, object_id)
            eq(self.ioctx.read(object_id), b'54321')
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'54789', 0)
            write_op.write(b'12345', 0)
            try:
                self.ioctx.operate_write_op(write_op, object_id)
            except Exception as e:
                logging.info("object content does not match")
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54321', 0)
            self.ioctx.operate_read_op(read_op, object_id)

```

@tchaikov  Is this Ok？

Reply:
```
def test_cmpext_op(self):
        object_id = 'test'
        with WriteOpCtx() as write_op:
            write_op.write(b'12345', 0)
            self.ioctx.operate_write_op(write_op, object_id)
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'12345', 0)
            write_op.write(b'54321', 0)
            self.ioctx.operate_write_op(write_op, object_id)
            eq(self.ioctx.read(object_id), b'54321')
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'54789', 0)
            write_op.write(b'12345', 0)
            try:
                self.ioctx.operate_write_op(write_op, object_id)
            except Exception as e:
                logging.info("object content does not match")
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54321', 0)
            self.ioctx.operate_read_op(read_op, object_id)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'12345', 0)
            try:
                self.ioctx.operate_read_op(read_op, object_id)
            except Exception as e:
                logging.info("object content does not match")
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)

```
@tchaikov Is this Ok？

Reply:
please note this is the restriction of your implementation, not the restriction of librados. there are at least two ways to address this:

1. to define a member variable of `WriteOp` using `cdef`, so you can pass the address of this member variable to `rados_write_op_cmpext()`.
1. or simply check the return value of `rados_write_op_operate()`. there are more than one way to do this. i'd suggest 
    1. derive a new class `ExtendMismatch` from `Error`. this class should have a property containing the offset of the first mismatched byte.
    2. extend `make_ex()` so any return value less than `-MAX_ERRNO` is used to create an instance of `ExtendMismatch`.

Reply:
```
import rbd
from rados import (Rados, WriteOp,  WriteOpCtx, ReadOpCtx, ObjectNotFound, MAX_ERRNO)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import rados
import unittest
import logging

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool_01')
        assert self.rados.pool_exists('pool_01')
        self.ioctx = self.rados.open_ioctx('pool_01')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool_01')
        self.rados.shutdown()

    def test_cmpext_op(self):
        object_id = 'test'
        with WriteOpCtx() as write_op:
            write_op.write(b'12345', 0)
            self.ioctx.operate_write_op(write_op, object_id)
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'12345', 0)
            write_op.write(b'54321', 0)
            self.ioctx.operate_write_op(write_op, object_id)
            eq(write_op.get_cmpext_result(), 0)
            eq(self.ioctx.read(object_id), b'54321')
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'56789', 0)
            write_op.write(b'12345', 0)
            try:
                self.ioctx.operate_write_op(write_op, object_id)
            except Exception as e:
                #the cmpext_result compare with expected error number, it should be (-MAX_ERRNO - 1)
                #where "2" is the offset of the first unmatched byte
                eq(write_op.get_cmpext_result(), -MAX_ERRNO-1)
                eq(self.ioctx.read(object_id), b'54321')
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54321', 0)
            self.ioctx.operate_read_op(read_op, object_id)
            eq(read_op.get_cmpext_result(), 0)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54789', 0)
            try:
                self.ioctx.operate_read_op(read_op, object_id)
            except Exception as e:
                #the cmpext_result compare with expected error number, it should be (-MAX_ERRNO - 2)
                #where "2" is the offset of the first unmatched byte
                eq(read_op.get_cmpext_result(), -MAX_ERRNO-2)
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        

if __name__ == '__main__':
    unittest.main()

```
```
root@jw-KVM:/data/ceph/build# python3 test_cmpext_op.py 
.
----------------------------------------------------------------------
Ran 1 test in 2.821s

OK

```
@tchaikov please check

Reply:
```
import rbd
from rados import (Rados, WriteOp,  WriteOpCtx, ReadOpCtx, ObjectNotFound, MAX_ERRNO)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import rados
import unittest
import logging

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool_01')
        assert self.rados.pool_exists('pool_01')
        self.ioctx = self.rados.open_ioctx('pool_01')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool_01')
        self.rados.shutdown()

    def test_cmpext_op(self):
        object_id = 'test'
        with WriteOpCtx() as write_op:
            write_op.write(b'12345', 0)
            self.ioctx.operate_write_op(write_op, object_id)
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'12345', 0)
            write_op.write(b'54321', 0)
            self.ioctx.operate_write_op(write_op, object_id)
            eq(write_op.get_cmpext_result(), 0)
            eq(self.ioctx.read(object_id), b'54321')
        with WriteOpCtx() as write_op:
            write_op.cmpext(b'56789', 0)
            write_op.write(b'12345', 0)
            try:
                self.ioctx.operate_write_op(write_op, object_id)
            except Exception as e:
                #the cmpext_result compare with expected error number, it should be (-MAX_ERRNO - 1)
                #where "1" is the offset of the first unmatched byte
                eq(write_op.get_cmpext_result(), -MAX_ERRNO-1)
                eq(self.ioctx.read(object_id), b'54321')
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54321', 0)
            self.ioctx.operate_read_op(read_op, object_id)
            eq(read_op.get_cmpext_result(), 0)
        with ReadOpCtx() as read_op:
            read_op.cmpext(b'54789', 0)
            try:
                self.ioctx.operate_read_op(read_op, object_id)
            except Exception as e:
                #the cmpext_result compare with expected error number, it should be (-MAX_ERRNO - 2)
                #where "2" is the offset of the first unmatched byte
                eq(read_op.get_cmpext_result(), -MAX_ERRNO-2)
            else:
                message = "cmpext did not raise Exception when object content does not match"
                raise AssertionError(message)
        

if __name__ == '__main__':
    unittest.main()

```
```
root@jw-KVM:/data/ceph/build# python3 test_cmpext_op.py 
.
----------------------------------------------------------------------
Ran 1 test in 2.821s

OK


```
@tchaikov please check

Reply:
@tchaikov updated, please review.

Reply:
@tchaikov updated, please review.